File: bot/src/voice/probe.ts
import { fetchJson } from "../util/http.js";
import { HttpTtsClient } from "./tts.js";

/** Probe a sidecar GET /health endpoint. */
export async function probeHttpHealth(baseUrl: string, timeoutMs = 5000): Promise<boolean> {
  const url = baseUrl.replace(/\/$/, "");
  try {
    const data = await fetchJson<{ ok?: boolean }>(`${url}/health`, { timeoutMs });
    return !!data?.ok;
  } catch {
    return false;
  }
}

/** Probe TTS with a one-word synthesis request. */
export async function probeHttpTts(
  baseUrl: string,
  voice: string,
  timeoutMs = 15000,
): Promise<boolean> {
  try {
    const client = new HttpTtsClient({ url: baseUrl, voice, timeoutMs });
    const { audio } = await client.synthesize("ok");
    return audio.length > 0;
  } catch {
    return false;
  }
}

/** @deprecated Use probeHttpTts */
export const probeKokoroTts = probeHttpTts;

/** Probe STT with a minimal non-silent PCM buffer. */
export async function probeHttpStt(baseUrl: string, timeoutMs = 10000): Promise<boolean> {
  const url = baseUrl.replace(/\/$/, "");
  try {
    const pcm = Buffer.alloc(256, 1);
    const data = await fetchJson<{ text?: string }>(`${url}/asr`, {
      method: "POST",
      timeoutMs,
      headers: {
        "Content-Type": "application/octet-stream",
        "X-Sample-Rate": "16000",
        "X-Channels": "1",
      },
      body: new Uint8Array(pcm),
    });
    return typeof data?.text === "string";
  } catch {
    return false;
  }
}

/** @deprecated Use probeHttpStt */
export const probeSherpaStt = probeHttpStt;

File: crates/mp-voice/src/probe.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

use std::time::Duration;

use serde::Deserialize;

use crate::tts::HttpTtsClient;

#[derive(Deserialize)]
struct HealthBody {
    ok: Option<bool>,
}

pub async fn probe_http_health(base_url: &str, timeout: Duration) -> bool {
    let url = format!("{}/health", base_url.trim_end_matches('/'));
    let Ok(client) = reqwest::Client::builder().timeout(timeout).build() else {
        return false;
    };
    match client.get(&url).send().await {
        Ok(res) => res
            .json::<HealthBody>()
            .await
            .ok()
            .and_then(|b| b.ok)
            .unwrap_or(false),
        Err(_) => false,
    }
}

pub async fn probe_http_stt(base_url: &str, timeout: Duration) -> bool {
    let url = format!("{}/asr", base_url.trim_end_matches('/'));
    let Ok(http) = reqwest::Client::builder().timeout(timeout).build() else {
        return false;
    };
    match http
        .post(&url)
        .header("content-type", "application/octet-stream")
        .header("X-Sample-Rate", "16000")
        .header("X-Channels", "1")
        .body(vec![1u8; 256])
        .send()
        .await
    {
        Ok(res) => res
            .json::<serde_json::Value>()
            .await
            .ok()
            .and_then(|v| v.get("text").map(|t| t.is_string()))
            .unwrap_or(false),
        Err(_) => false,
    }
}

pub async fn probe_http_tts(base_url: &str, voice: &str, timeout: Duration) -> bool {
    let _ = timeout;
    let client = HttpTtsClient::new(base_url, voice);
    match client.synthesize("ok").await {
        Ok((audio, _)) => !audio.is_empty(),
        Err(_) => false,
    }
}
